openedx/core/lib/graph_traversals.py

```diff
@@ -0,0 +1,257 @@
+
+    Yields:
+        get_result(node): The result of the next node in the traversal.
+    """
+    return _traverse_generic(
+        start_node,
+        get_parents=get_parents,
+        get_children=get_children,
+        get_result=get_result,
+        filter_func=filter_func,
+        yield_descendants_of_unyielded=yield_descendants_of_unyielded,
+    )
+
+
+def traverse_pre_order(
+    start_node, get_children, get_result=None, filter_func=None,
+):
+    """
+    Generator for yielding nodes of a tree (or directed acyclic graph)
+    in a pre-order sort.
+
+    Arguments:
+        See description in traverse_topologically.
+"""
+    return _traverse_generic(
+        start_node,
+        # There's no need to get_parents
+        get_parents=None,
+        get_children=get_children,
+        get_result=get_result,
+        filter_func=filter_func,
+    )
+
+
+def traverse_post_order(
+    start_node, get_children, get_result=None, filter_func=None
+):
+    """
+    Generator for yielding nodes of a tree (or directed acyclic graph)
+    in a post-order sort.
+
+    Arguments:
+        See description in traverse_topologically.
+    """
+    class _Node(object):
+        """
+        Wrapper class for nodes that keeps a children iterator.
+        """
+        def __init__(self, node, get_children):
+            self.node = node
+            self.children = iter(get_children(node))
+
+    # If get_result or filter_func aren't provided, make them no-ops.
+    get_result = get_result or (lambda node_: node_)
+    filter_func = filter_func or (lambda __: True)
+
+    # Use deque for the stack, which is O(1) for pop and append.
+    # Use the _Node class to keep track of iterated children.
+    stack = deque([_Node(start_node, get_children)])
+
+    # Keep track of which nodes have been visited.
+    visited = set()
+
+    while stack:
+        # Peek at the current node at the top of the stack.
+        current = stack[len(stack) - 1]
+
+        # Verify the node wasn't already visited and the node
+        # satisfies the filter_func.
+        if current.node in visited or not filter_func(current.node):
+            # Since already visited or filtered out, remove from the
+            # stack and continue with the next node.
+            stack.pop()
+            continue
+
+        # See if there are any additional children for this node.
+        try:
+            next_child = current.children.next()
+
+        except StopIteration:
+            # Since there are no children left, visit the node and
+            # remove it from the stack.
+            yield get_result(current.node)
+            visited.add(current.node)
+            stack.pop()
+
+        else:
+            # If so, add the child to the top of the stack.
+            stack.append(_Node(next_child, get_children))
+
+
+def _traverse_generic(
+    start_node, get_parents, get_children, get_result=None, filter_func=None, yield_descendants_of_unyielded=False,
+):
```

Comment:
Pylint wants L176 indented 8 spaces.  Probably cleanest to either indent L177 4 spaces, or put it at the end of L176.